Pkgconf recursive attr path resolution

fbudin69500 opened the pull request.

fbudin69500 commented:
@mwoehlke : This patch does not yet include a `cps2pc` conversion script. This could be added in a different patch, or in this patch if you think it is better.

mwoehlke-kitware commented:
Pedantic: in your last commit, "path in which the library is searches for" presumably should be "searche**d**"?

fbudin69500 commented:
@mwoehlke good catch!

fbudin69500 commented:
@mwoehlke Do you need me to improve/modify anything else?

mwoehlke-kitware commented on pc2cps:
```diff
@@ -9,7 +9,8 @@ import shlex
 
-#==============================================================================
+
+# ==============================================================================
```
Please delete the terminal character to keep these at 79 characters.

TBH, I am not a fan of PEP-8. I particularly loath the requirement for two empty lines, because it is very difficult to remember when to use two and when to use only one, and it's also completely unnecessary when using separator lines (which are much better for legibility anyway than just whitespace). I would prefer disabling that particular check.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -201,8 +201,8 @@ def main(args):
         'Default-Components': [':%s' % component_name],
     }
     extend(spec, 'Version', package.attr.get('Version'))
-    extend(spec, 'Description', package.attr.get('Description'))
```
Both `Description` and `Website` are part of the [Supplemental Schema](https://mwoehlke.github.io/cps/schema-supplement.html). It may be that the reader needs to preserve and allow access to unknown fields in addition to explicit extensions, but replacing these with pkgconfig-specific extensions is the wrong thing to do.

fbudin69500 replied:
How is the supplemental schema used in `cps.py`? When I read a CPS file that contains these fields, they seem to be ignored and discarded. I don't see them in the `components` variable.

mwoehlke-kitware replied:
It isn't, presently. Hence my comment: "it may be that the reader needs to preserve and allow access to unknown fields in addition to explicit extensions"...

I haven't thought about what form that should take; considering we would be adding such a feature to handle stuff from the supplemental schema, it probably should not be overly obtuse.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -36,6 +50,7 @@ class pkgconfig(object):
                 if ma is not None:
                     self.attr[ma.group(1)] = self._expand(ma.group(2))
                     continue
+        self._extract_requires()
```
Blank line before this, maybe?

I notice your later code is also whitespace impoverished. Please consider adding some blank lines to help separate blocks of code into logical sections.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -12,18 +12,32 @@ import sys
+             'a': 'archive',
+             None: 'Interface'
+            }
     #--------------------------------------------------------------------------
```
Missing blank line before function.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,18 +86,60 @@ class pkgconfig(object):
-        extend(c, 'Link-Flags', self._get_split('Libs'))
+            return [] if x is None else x.keys()
+
+        def _list(re_):
```
I'm not in love with names ending in `_`; maybe use `pattern` instead? Also, it is not obvious from the name what purpose this function serves... and looking at how it is used, I wonder if it wouldn't be better to replace uses with (e.g.):

```python
  [x[2:] for x in link_flags if x.startswith('-L')]
```

(...which is also probably faster than using regular expressions...)

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,18 +86,60 @@ class pkgconfig(object):
+        # If not found and if package name given in library list with
+        # "-l" flag, an error should be thrown?
+        def _find_location():
+            for path in self._paths:
```
Does `_paths` need to be a member variable? It seems like a local would be okay?

fbudin69500 replied:
Indeed.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,18 +86,60 @@ class pkgconfig(object):
+        self._paths = [x for x in _list(self._re_libpath) if os.path.isdir(x)]
+        # Find all libraries
+        self.libname = _list(self._re_libname)
+        c = {}
```
Please give this a more descriptive name

fbudin69500 replied:
I renamed it to `_component` (since there is already a method named `component`).

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,18 +86,60 @@ class pkgconfig(object):
+                    pattern = '(lib)?' + self.library_name + self._lib_pattern
+                    ma = re.match(pattern, file)
+                    if ma is not None:
+                        return os.path.join(path, ma.group(0))
```
I think this is missing logic to deal with the case of both shared and static libraries matching. (Um... to *properly* handle that would require being aware of the linker flags that control which library the linker will choose in such cases... maybe we don't need to go *that* far, although a `TODO` might be in order, but it seems like we should at least implement our own preference, rather than relying on the order in which files are iterated?)

Actually, I wonder if it wouldn't be better to approach this entire problem as a state machine; that is, parse each link flag one at a time and take appropriate action, which would be either modifying the parse state, locating a library to add to the list of link libraries, or passing along a link flag. (This also makes me wonder if we ever want to pass along the link flags to change the linker's static/shared preference...)

fbudin69500 replied:
Indeed, I could add a flag to specified the preferred order in which the type of the library (static/shared) is chosen.

At first, when I started to work on this PR, I was searching for all the libraries that were in the linker flags. But since `pc` files are, as far as I know, only declare one component/target, I thought it would make sense to convert the libraries found in the flags to `required` components instead of looking for their paths.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,18 +86,60 @@ class pkgconfig(object):
+        # some library names already contain the prefix `lib`.
+        self.libname = [x for x in self.libname if not re.match('(lib)?' + x, self.library_name)]
+        # Update requires dictionary
+        self.requires.update({x.lower(): {} for x in self.libname if x.lower()
```
This is a little awkward; I think it would be better to break before the `if`, rather than in the middle of the conditional expression

mwoehlke-kitware commented on pc2cps:
```diff
@@ -132,10 +188,11 @@ def main(args):
     parser.add_argument(
         'input', metavar='PACKAGE', type=str,
         help='Input pkg-config (.pc) file name or package name')
+    parser.add_argument("-n", "--library-name", required=False,
```
The change in line wrapping style here vs. the previous line is a little jarring, as is the use of `"` for strings, which I don't see happening elsewhere. Also, the `.` at the end of the `help` is inconsistent.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,18 +86,60 @@ class pkgconfig(object):
+            return [] if x is None else x.keys()
+
+        def _list(re_):
+            ma = [re_.match(x) for x in link_flags]
```
Please use `mf` here instead of `ma`; → Match Flag. (Or just use `m` if there is no need to differentiate between different matches.) Similar for other instances of `ma`.

mwoehlke-kitware commented:
I believe the only reason 2172ebe is not merged already is because, immediately after it was written, it was not obvious that anyone cared, and it thus stalled. If we now have a definite goal of integrating this functionality, could we merge that first so that it's easier to review the later improvements?

Also, another note on making PEP-8 related changes, if we're going to do *any* of that, I would strongly prefer that it be all-or-nothing, i.e. either apply the changes to the entire repo, or continue to follow the existing style.

fbudin69500 commented:
Thanks for the review! The PEP-8 commit can be ignored if you want. It doesn't really matter to me, I was just trying to follow the Python rules. In the end, Python can be very inflexible, and I do like flexibility...
I will address the rest of the comments and update the commits.

I think people are finally interested in this functionality, so you can merge your current branch, and I will cherry-pick my commit onto a new branch based on the new master.

mwoehlke-kitware commented:
> The PEP-8 commit can be ignored if you want.

I'm not adverse to following those rules that are useful, though as mentioned I'd rather make such changes across the repository, for consistency, rather than to just one file. (Possibly you've noticed already, but I'm basically okay with all the rules *except* the one about multiple blank lines, and I'm lukewarm on the one about rigidly having a space after comments. The latter makes sense for *real* comments, but IMHO isn't necessary for separator lines.)

Let's please drop it from *this* PR (but do please squash the line length fix into the previous commit :slightly_smiling_face:). If you like, I'd be happy to revisit in a separate PR, but I think it would be best to keep it separate.

fbudin69500 commented:
@mwoehlke Let me know if this new version of the PR can be rebased onto the new master once you have merged the `pycps` branch.

fbudin69500 commented:
@mwoehlke-kitware @mwoehlke ping @jamiesnape

jamiesnape commented:
Indeed ping @mwoehlke-kitware @mwoehlke, we need this ASAP please.

mwoehlke-kitware commented:
Please rebase. I also went ahead and took the first of your two commits...

Commit:
Convert compilation flags to CPS attributes

`-l` and `-L` flags are recognized and values given through these
flags are converted to known CPS attributes, `Requires` and `Location`.

The value of the `Location` attribute is computed using the information
contained in `-L` (path in which the library is searched for) and
the library name. The library name can be given with a new command line
option `--library-name`. If it is not specified, it will be assumed that
the name of the library is the base name of the input pkgconfig file.

The flags `-l` and `-L` are removed from the list of flags saved under
`Link-Flags` except if some libraries are marked as black-listed.

The library searched for in the path given by the flags `-L` is assumed
to have one of the following file extension: `.a`, `.so`, `.dylib`. If
it is a shared object, it can also contain a version number.

fbudin69500 commented:
I rebased this branch onto master and improved its style to match the style of the rest of the script.

fbudin69500 commented:
@mwoehlke @mwoehlke-kitware @jamiesnape ping

jamiesnape commented:
@mwoehlke @mwoehlke-kitware @fbudin69500 ping.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -113,11 +182,11 @@ def parse(path_or_name):
 
 #------------------------------------------------------------------------------
 def extend(data, key, value):
-    if value is not None:
```
What is the purpose of this change?

fbudin69500 replied:
Just making the code simpler. I can revert if you really dislike it. I just don't see the point of adding 3 words here.

mwoehlke-kitware replied:
Upon further poking, I see it is also useful for suppressing empty lists or maps. However, as-is, this change is potentially broken, since values like `False`, `0`, that *should* be set, aren't. (Maybe that doesn't matter now, but for a generic function, I'd prefer to not have it be broken for certain inputs, even if we never give it those inputs at present.)

I have a better version locally. Please hold off on edits pending resolution of the latest general discussion.

fbudin69500 replied:
Ok, just let me know when you have merged your local version. I will rebased on the latest master then.

mwoehlke-kitware commented on pc2cps:
```diff
@@ -72,27 +92,76 @@ class pkgconfig(object):
+        # Add system libraries found in link flags to requirements
+        list_requires = _keys(self.requires)
+        # Update requires dictionary
+        self.requires.update({x.lower(): {} for x in self.libname if x.lower()
```
Upon further consideration, I do believe forcibly lower-casing these is broken. Package names, in both CPS and pkg-config, are not necessarily lower case, and CPS is (partially) case-sensitive.

fbudin69500 replied:
True. I had only seen packages with lower case but looking more I see that some pkg files do have uppercase. I'll update this.

mwoehlke-kitware commented:
TBH, while I think it makes sense to try to resolve the library name for the `.pc` being processed, I am concerned that trying to translate *other* link flags into requirements is doomed to failure. Just poking around a bit in what `.pc` files I have lying around, I see a lot of instances of `-lz`, `-lm`, `-ldl` and so forth, *none* of which match any other `.pc` files.

In fact, zlib is a great, pathological example. Running this version of `pc2cps` on `zlib.pc` not only fails to find the `Location` of the library, but adds a dependence on a non-existing `z` package (which, of course, *should* be the same CPS we are generating!).

What is the problem with the link flags?

fbudin69500 commented:
@mwoehlke : The goal of replacing the link flags with the requirements is to make the cps file clearer and more explicit. Potentially a `pc` file could be converted to `cps` and used on a different platform that or with a compiler that uses different flags.
I anticipated the issue you are mentioning and that is why there is the flag `-n` that allows the user to specify the name of the library if necessary. You can use the following command line and then everything works:
```
./pc2cps /usr/lib/x86_64-linux-gnu/pkgconfig/zlib.pc -n z                                                                             {
  "Components": {
    "zlib": {
      "Includes": [
        "/usr/include"
      ], 
      "Location": "/usr/lib/x86_64-linux-gnu/libz.so", 
      "Type": "dylib"
    }
  }, 
  "Cps-Version": "0.8.0", 
  "Default-Components": [
    ":zlib"
  ], 
  "Description": "zlib compression library", 
  "Name": "zlib", 
  "Version": "1.2.8"
}
```

mwoehlke-kitware commented:
> Potentially a `pc` file could be converted to `cps` and used on a different platform

Given the extent to which this PR adds a reliance on the libraries actually existing on the filesystem when `pc2cps` is run, I find that reasoning... a bit implausible.

> I anticipated the issue you are mentioning and that is why there is the flag -n that allows the user to specify the name of the library if necessary.

Okay, that helps for at least not creating pathologically recursive CPS's.

What about creating requirements on packages that don't exist? (e.g. `z`, `m`, `dl`) Are we basically just accepting that the output may require hand editing? I'm starting to think I'd feel better if that part of the conversion was optional. Would that be acceptable?

fbudin69500 commented:
Yes, that's sounds like a good idea to leave the requirements optional.

fbudin69500 commented:
@mwoehlke I added both a option to keep the flags as is, and to be able to blacklist some libraries, such as `m`.

mwoehlke-kitware commented:
Does ~~bd5e32274a4c~~ 2e3f53379c28 work for you?

mwoehlke-kitware commented:
BTW, apologies for my take not following your most recent changes more closely; I'd done most of the work before I saw your latest updates, so wound up reimplementing your changes in what I already had rather than starting from scratch.

fbudin69500 commented:
No problem. I found one default value missing, but otherwise it looks good.
Thanks!

On Wed, Jun 20, 2018 at 5:08 PM Matthew Woehlke <notifications@github.com>
wrote:

> BTW, apologies for my take not following your most recent changes more
> closely; I'd done most of the work before I saw your latest updates, so
> wound up reimplementing your changes in what I already had rather than
> starting from scratch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mwoehlke/pycps/pull/6#issuecomment-398897523>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAYZGOmMZZSTCIwPLqFVlCpliHzSK4_Tks5t-rnhgaJpZM4UALZn>
> .
>

mwoehlke-kitware commented:
Okay, take 3... does cf462f95fa7c work? If yes, I will promote that to master and close this.

Thanks!

fbudin69500 commented:
Looks good. Thanks!

mwoehlke commented:
Integrated as cf462f95fa7c.

fbudin69500 commented:
@mwoehlke Thanks. Out of the 3 other PRs, 2 should be pretty straightforward.
